cmd/nerdctl/container_diff.go

```diff
@@ -0,0 +1,225 @@
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/leases"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/platforms"
+	"github.com/containerd/continuity/fs"
+	"github.com/containerd/log"
+	"github.com/containerd/nerdctl/pkg/api/types"
+	"github.com/containerd/nerdctl/pkg/clientutil"
+	"github.com/containerd/nerdctl/pkg/idgen"
+	"github.com/containerd/nerdctl/pkg/idutil/containerwalker"
+	"github.com/containerd/nerdctl/pkg/imgutil"
+	"github.com/containerd/nerdctl/pkg/labels"
+	"github.com/opencontainers/image-spec/identity"
+	"github.com/spf13/cobra"
+)
+
+func newDiffCommand() *cobra.Command {
+	var diffCommand = &cobra.Command{
+		Use:               "diff [flags] [CONTAINER]",
+		Short:             "Inspect changes to files or directories on a container's filesystem",
+		Args:              cobra.MinimumNArgs(1),
+		RunE:              diffAction,
+		ValidArgsFunction: diffShellComplete,
+		SilenceUsage:      true,
+		SilenceErrors:     true,
+	}
+	return diffCommand
+}
+
+func processContainerDiffOptions(cmd *cobra.Command) (types.ContainerDiffOptions, error) {
+	globalOptions, err := processRootCmdFlags(cmd)
+	if err != nil {
+		return types.ContainerDiffOptions{}, err
+	}
+
+	return types.ContainerDiffOptions{
+		Stdout:   cmd.OutOrStdout(),
+		GOptions: globalOptions,
+	}, nil
+}
+
+func diffAction(cmd *cobra.Command, args []string) error {
+	options, err := processContainerDiffOptions(cmd)
+	if err != nil {
+		return err
+	}
+
+	client, ctx, cancel, err := clientutil.NewClient(cmd.Context(), options.GOptions.Namespace, options.GOptions.Address)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	walker := &containerwalker.ContainerWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found containerwalker.Found) error {
+			if found.MatchCount > 1 {
+				return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
+			}
+			changes, err := getChanges(ctx, client, found.Container)
+			if err != nil {
+				return err
+			}
+
+			for _, change := range changes {
+				switch change.Kind {
+				case fs.ChangeKindAdd:
+					fmt.Fprintln(cmd.OutOrStdout(), "A", change.Path)
+				case fs.ChangeKindModify:
+					fmt.Fprintln(cmd.OutOrStdout(), "C", change.Path)
+				case fs.ChangeKindDelete:
+					fmt.Fprintln(cmd.OutOrStdout(), "D", change.Path)
```

Comment:
OutOrStdout should be called only on e

Reply:
Fixed at https://github.com/containerd/nerdctl/pull/2626/commits/3d1a5f1157be20bcd801181d1582e8edc074df60
Thanks a lot really. 👍